core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -311,6 +313,14 @@ public Node start() {
         // Start the transport service now so the publish address will be added to the local disco node in ClusterService
         TransportService transportService = injector.getInstance(TransportService.class);
         transportService.start();
+        DiscoveryNode localNode = injector.getInstance(DiscoveryNodeService.class)
+                .buildLocalNode(transportService.boundAddress().publishAddress());
+        clusterService.setLocalNode(localNode);
+        transportService.setLocalNode(localNode);
+        clusterService.add(transportService.getTaskManager());
+
+        // state NodeConnectionService last, so it's listeners will be first and last in cluster service
```

Comment:
that sucks, sorry :) We can't build stuff that is dependent on the order of how the listeners are added! Can we find a better way of doing this? I think each listener needs to have priority or so and every prioritoy can only be added once? Maybe we don't need this to be a list at all? This stuff is so fragile we have to iterate until it's safe